dot/state/epoch.go

```diff
@@ -220,8 +240,30 @@ func (s *EpochState) SetEpochData(epoch uint64, info *types.EpochData) error {
 	return s.db.Put(epochDataKey(epoch), enc)
 }
 
-// GetEpochData returns the epoch data for a given epoch
-func (s *EpochState) GetEpochData(epoch uint64) (*types.EpochData, error) {
+// GetEpochData returns the epoch data for a given epoch persisted in database
+// otherwise will try to get the data from the in-memory map using the header
+// if the header params is nil then it will search only in database
+func (s *EpochState) GetEpochData(epoch uint64, header *types.Header) (*types.EpochData, error) {
+	epochData, err := s.GetEpochDataInDatabase(epoch)
+	if err == nil && epochData != nil {
+		return epochData, nil
+	}
+
+	// if no header is given then skip the lookup in-memory
+	if !errors.Is(err, chaindb.ErrKeyNotFound) || header == nil {
+		return nil, fmt.Errorf("failed to get epoch data from database: %w", err)
```

Comment:
If the error is `nil` and the header is nil as well, the error will be `nil` and the error wrap will not wrap properly. Maybe have these two conditions as two separate if conditions?

Reply:
good catch!!